Spec: Add SigV4 Auth Support for Catalog Federation

XJDKC opened the pull request:
<!--
    Possible security vulnerabilities: STOP here and contact security@apache.org instead!

    Please update the title of the PR with a meaningful message - do not leave it "empty" or "generated"
    Please update this summary field:

    The summary should cover these topics, if applicable:
    * the motivation for the change
    * a description of the status quo, for example the current behavior
    * the desired behavior
    * etc

    PR checklist:
    - Do a self-review of your code before opening a pull request
    - Make sure that there's good test coverage for the changes included in this PR
    - Run tests locally before pushing a PR (./gradlew check)
    - Code should have comments where applicable. Particularly hard-to-understand
      areas deserve good in-line documentation.
    - Include changes and enhancements to the documentation (in site/content/in-dev/unreleased)
    - For Work In Progress Pull Requests, please use the Draft PR feature.

    Make sure to add the information BELOW this comment.
    Everything in this comment will NOT be added to the PR description.
-->

## Context

### Prior Catalog Federation work:
* https://github.com/apache/polaris/pull/1026
* https://github.com/apache/polaris/pull/1305


### Reference

Below are all the reference mentioned in this PR:

* AWS Docs:
  * AWS SigV4 protocol: https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-authenticating-requests.html
  * Configure IAM for Glue: https://docs.aws.amazon.com/glue/latest/dg/configure-iam-for-glue.html
  * [AWS STS Regions and endpoints](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_temp_region-endpoints.html)
* Iceberg SDK 
  * [AuthProperties.java#30](https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/rest/auth/AuthProperties.java#L30)
  * [RESTSigV4Signer.java#L91](https://github.com/apache/iceberg/blob/apache-iceberg-1.9.0/aws/src/main/java/org/apache/iceberg/aws/RESTSigV4Signer.java#L91)
  * [AwsProperties.java#L417-L435](https://github.com/apache/iceberg/blob/apache-iceberg-1.9.0/aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java#L417-L435)
  * [RESTSigV4AuthSession.java#L71](https://github.com/apache/iceberg/blob/07ad9a635c6e24959280f53b76ef3036e33d13ee/aws/src/main/java/org/apache/iceberg/aws/RESTSigV4AuthSession.java#L71)
* Polaris Code
  * param to skip assuming IAM role: [SKIP_CREDENTIAL_SUBSCOPING_INDIRECTION](https://github.com/apache/polaris/blob/4db7998381a61e9cab82cdc4fded6867b0bca464/service/common/src/main/java/org/apache/polaris/service/catalog/io/FileIOUtil.java#L92)
  * roleSessionName of the StorageConfig: https://github.com/apache/polaris/blob/ca35bbaf10aff807053bfc032f8e71a3f007cc79/polaris-core/src/main/java/org/apache/polaris/core/storage/aws/AwsCredentialsStorageIntegration.java#L68

## Description
This PR adds the SigV4 Auth Support so that Polaris can federate to Glue IRC / S3Tables / catalog server hosted behind the AWS API Gateway.

This PR is based on this new proposal for creds management in Polaris:

[Apache Polaris Creds Management Proposal](https://docs.google.com/document/d/1MAW87DtyHWPPNIEkUCRVUKBGjhh5bPn0GbtV7fifm30/edit?usp=sharing)

## Details
User-provided properties:
* `roleArn`: the AWS IAM role urn, polaris will assume the role to get a tmp aws session credential
* `roleSessionName`: The role session name to be used by the SigV4 protocol for signing requests
* `externalId`: An optional external id used to establish a trust relationship with AWS in the trust policy
* `signingRegion`: Region to be used by the SigV4 protocol for signing requests
* `signingName`: The service name to be used by the SigV4 protocol for signing requests, the default signing name is "execute-api" is if not provided
    * AWS Glue: `glue`
    * AWS S3Tables: `s3tables`
    * AWS API Gateway: `execute-api`

Service-provided properties:
* `serviceIdentity`: it's a nested object. In this PR, I only added two identity types:
    * **AWS_IAM_USER**: Polaris is configured with long-lived AWS credentials from an IAM user. It uses these credentials to assume the IAM role provided by Polaris users, then accesses the remote catalog using the subscoped temp credentials.
      * `userArn`: The aws user arn used to assume the aws role, this represents the polaris service itself, overall steps:
          * **Step 1**: Polaris will use an user aws credential to assume the IAM role, this aws credential is configured for polaris, usually polaris will read it from environment variable
          * **Step 2**: After assuming the role, polaris will get the tmp aws credential, polaris will put these credentials into the prop map, this prop map will be used to initialize the REST Catalog Client, properties can be found in [AwsProperties](https://github.com/apache/iceberg/blob/apache-iceberg-1.9.0/aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java#L174-L181): 
              * `rest.signing-region`
              * `rest.signing-name`
          * **Step 3**: [RESTSigV4Signer](https://github.com/apache/iceberg/blob/apache-iceberg-1.9.0/aws/src/main/java/org/apache/iceberg/aws/RESTSigV4Signer.java) will sign the requests based on SigV4 protocol
    * **AWS_IAM_ROLE**: Polaris retrieves temporary credentials via [Amazon ECS container credentials](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-iam-roles.html) or [Amazon EC2 instance IAM role credentials](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html), then assumes the polaris user-provided IAM role to get new temporary credentials to access the remote catalog.

dimas-b commented on spec/polaris-management-service.yml:
```diff
@@ -938,6 +940,34 @@ components:
+      properties:
+        roleArn:
+          type: string
+          description: The aws IAM role arn assume when signing requests
```
Does this assume the use of STS? 

Mere SigV4 can be done with just a keyId/keySecret pair.. why is it not an option? AFAIK, there are S3 implementations that do not have STS, but support SigV4.

XJDKC replied:
Yes, this assumes the use of STS. While SigV4 can technically work with just a keyID/keySecret pair, that's not how it works in Polaris.

Let me break it down a bit:

Polaris acts as the service provider, and it has its own IAM user with an AWS credential (key ID and key secret). However, this IAM user is owned by the Polaris service, not by the Polaris user. Since IAM doesn't allow one AWS account to grant privileges directly to another IAM user belong to another AWS account, Polaris wouldn't be able to access the polaris user's Glue catalog that way.

So how does Polaris access a user's Glue catalog? By assuming the IAM role provided by the Polaris user. This lets the Polaris service temporarily inherit the permissions tied to that role, essentially gaining the necessary access without the user having to expose long-lived credentials.

Now, why are there S3 implementations that don't use STS?
That's very common on the query engine side. In that case, the query engine is fully managed by the Polaris user themselves. They can create an IAM user + access key pair, grant that IAM user privileges to their storage, and configure their engine to use that credential directly. No need for STS in that scenario.

So here's the key difference:
1. AWS User credentials (key ID/secret) are long-lived and tied to a user. IAM roles aren’t real credentials—they’re just a set of permissions. To access resources, Polaris assumes a role and obtains short-lived temporary credentials via STS. This is much more secure.
2. In the query engine use case, both the engine and the storage are owned by the same identity (the Polaris user), so they're free to use long-lived user credentials if they want.
3. The IAM role assumption model is actually consistent with how Polaris accesses S3 storage in general. In Polaris S3 storage configs, users provide an IAM role, Polaris assumes that role, and gets subscoped temporary credentials via STS to access the storage.

dimas-b replied:
Thanks for the info. Re: `Polaris acts as the service provider ...` while I agree that it's a possible use case, I do not think Apache Polaris needs to be restricted to just that use case. I can imagine users who'd want to deploy and use Polaris in a controlled environment and have ownership both of the AWS storage and Polaris servers.

That, of course, does not invalidate the security best practices that you referenced.

Given that the new connection config is apparently meant to enable connections to AWS services and not to generic S3 implementations, I'd propose to rename the config type to something like `AWSSTSAuthenticationParameters` to avoid ambiguity (with type code `AWSSTS`). WDYT?

XJDKC replied:
I think we should keep it consistent with the Iceberg SDK: [AuthProperties.java#30](https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/rest/auth/AuthProperties.java#L30)

Also, getting the tmp AWS credential from STS is just the first step. The actual operation involves signing the request using the SigV4 protocol. Given that, I'd prefer we stick with using SIGV4 as the type name, it better reflects the authentication mechanism being used. WDYT?

dimas-b replied:
To clarify my point on naming: indeed, the actual authentication mechanism on the wire is going to be SigV4, however, the user-facing configuration and requirements involve STS and role assumption. This is more specific than SigV4. In other words, users who configure this authentication method have to be aware of STS and roles. The basic inputs into SigV4 are ID/secret/session-token, but this auth config is higher level.

Therefore, from my POV using a name that highlights reliance on higher-level concepts like STS would be beneficial and provide clarity to users.

I do not feel too strongly about this, though :)

XJDKC replied:
Or it's better to have a subtype for SigV4? e.g. `STS` or `STATIC_CREDS`? WDYT?
This will cover both scenarios.

dennishuo replied:
Replied on the dev list as well, but I think for STATIC_CREDS we can probably follow the same convention we started with  [SKIP_CREDENTIAL_SUBSCOPING_INDIRECTION](https://github.com/apache/polaris/blob/4db7998381a61e9cab82cdc4fded6867b0bca464/service/common/src/main/java/org/apache/polaris/service/catalog/io/FileIOUtil.java#L92) for StorageConfig, where we can be more opinionated about subscoping design for the "managed multi-tenant" use cases, while allowing a simpler approach for open-ended "self-managed, single-tenant" use cases to just delegate to environment variables and/or default client SDK behaviors.

In particular, we can take stand on whether we ever actually want Polaris to run in a mode that mixes some catalogs that do subscoping with other catalogs that just use static credentials through environment variables or similar. The problem is that the subscoping is trying to promise certain guarantees of isolation, and mixing the two might fundamentally violate its assumptions. So it might be best to really bifurcate the two use cases by essentially making the behavior choice a top-level server config (e.g. SKIP_CREDENTIAL_SUBSCOPING_INDIRECTION) instead of trying to express it as an "option" in the ConnectionConfigType.

That said, I don't feel too strongly whether to make the auth type more sts-specific since *in theory* there could still be other non-static-creds based flows for sigv4 that also don't use STS. Like it could be `SIGV4_STS` or `STS_SIGV4`. But for now still not adding any particular type for static creds.

Alternatively, if anyone feels strongly that even in the static creds case we need an explicit ConnectionType for it (maybe in addition to a server-level config), we at the very least need to add a `SUPPORTED_CONNECTION_AUTH_TYPES` just like how we have a `SUPPORTED_CATALOG_STORAGE_TYPES` because in multi-tenant configurations it'll be important to block the ability for catalogs to be configured to use open-ended static creds

XJDKC replied:
Hey Dennis, thanks for bringing this up, I didn't realize there was already a solution in place for storage config.

We can definitely follow the same pattern for connection config: skip assuming the role and just read the credentials from the command line. I also looked into the Iceberg SDK, if we don't explicitly pass credentials in the catalog properties when initializing the RESTCatalog, the SDK falls back to AWS SDK's `DefaultCredentialsProvider`, which picks up credentials from the environment variables.

Code pointers:
[RESTSigV4Signer.java#L91](https://github.com/apache/iceberg/blob/apache-iceberg-1.9.0/aws/src/main/java/org/apache/iceberg/aws/RESTSigV4Signer.java#L91)
[AwsProperties.java#L417-L435](https://github.com/apache/iceberg/blob/apache-iceberg-1.9.0/aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java#L417-L435)

Agree that there could be other non-static-creds (e.g. For Top Secret Region, we need to call CAP service to get the credentials, it's a replacement to the STS). Instead of creating many top-level types for SigV4, let's not adding any particular type for SigV4 auth.

I will add `SUPPORTED_CONNECTION_AUTH_TYPES` in a follow up PR.

singhpk234 commented on spec/polaris-management-service.yml:
```diff
@@ -938,6 +940,34 @@ components:
           format: password
           description: Bearer token (input-only)
 
+    SigV4AuthenticationParameters:
```
should we also add an optional role-session-name for tracking this STS assume role ?

singhpk234 replied:
Side note how do infer STS endpoint ? do we just rely on SDK ? 
asking because there is a seperate FIPS endpoint for some of the regions

dimas-b replied:
Good point about STS endpoint. In this form of connection config, STS params are out of scope, I guess they are implied in the environment... but I think we have to support non-AWS storage at some point. Heads up: I'm thinking about opening a PR for STS endpoint for the normal storage access path (not Federated Catalogs).

singhpk234 replied:
It is a global but it works for  regions **_enabled by default_** not all regions here is the list of region it works for by default 
https://docs.aws.amazon.com/accounts/latest/reference/manage-acct-regions.html i.e why there is **_We recommend that you use the appropriate AWS STS regional endpoints_** 

Though I agree with @dimas-b that they need to be configured at a different place rather than here, may be via Polaris configs ? I am not sure if this today, we had a to deal with a lot of head-aches in EMR-FS for this. Hence brought out if we are onboarding a new use case on STS we should incorporate this.

Not a blocker for this pr !

dimas-b replied:
@XJDKC : your comment seems to assume that STS is provided by AWS. However, MinIO also have an STS implementation.

XJDKC replied:
Hey Prashant and Dmitri,

Even though AWS STS doesn't have regional endpoints in every region, I believe the AWS SDK will fall back to the global endpoint if a regional one isn't available. From what I understand in the AWS docs, you can opt out of using regional endpoints (though it’s not recommended), and if you don't opt out, the SDK will choose the appropriate endpoint: i.e. using the regional one when available, and falling back to the global endpoint otherwise.

I also took a look into MinIO: https://min.io/docs/minio/linux/developers/security-token-service.html
Since MinIO provides an S3-compatible API, if we want to support its STS service, Polaris will first need to support the s3compat API for the storage.

In addition, since we need to send requests to STS to get subscoped storage credentials, we'd need to define a unified approach that works across both the storage config and the connection config.

Given all of that, I'd recommend marking this out of scope for the initial spec changes. It's not a one-way door, we can always introduce the optional STS endpoint later once we've had time to fully think through the design.

XJDKC replied:
Also took a look into the `AwsCredentialsStorageIntegration`, seems that we use a fixed name. I can add an optional `roleSessionName` for connection config.

https://github.com/apache/polaris/blob/ca35bbaf10aff807053bfc032f8e71a3f007cc79/polaris-core/src/main/java/org/apache/polaris/core/storage/aws/AwsCredentialsStorageIntegration.java#L68

XJDKC replied:
Added an optional `roleSessionName` property to the `SigV4AuthenticationParameters`

dimas-b replied:
optional roleSessionName SGTM :+1:

dimas-b replied:
(off topic) Re: MinIO, it is indeed not supported now, but it was requested by users :) I'm working (gradually) on supporting it.

dimas-b replied:
Yes, I agree we need a unified approach for STS connections in direct storage access, sub-scoping credentials and catalog federation.

singhpk234 commented on spec/polaris-management-service.yml:
```diff
@@ -938,6 +940,34 @@ components:
+      properties:
+        roleArn:
+          type: string
+          description: The aws IAM role arn assume when signing requests
```
[optional] would it be better to add few words on who is assuming this role ? just for readability

XJDKC replied:
Ack!

singhpk234 commented on spec/polaris-management-service.yml:
```diff
@@ -938,6 +940,34 @@ components:
+        externalId:
+          type: string
+          description: An optional external id used to establish a trust relationship with AWS in the trust policy
+        signingRegion:
+          type: string
+          description: Region to be used by the SigV4 protocol for signing requests
```
does this region needs to be same as that of Polaris is running is Sigv4 suffciient for cross region or we need SigV4A ?

singhpk234 replied:
I see, my question is a bit different, do we need the assertion that AWS region of polaris service and signing region is same ? or do current SigV4 is can handle cross regions.

XJDKC replied:
Yeah, signing region is mandatory. If it's not explicitly provided, we can infer it based on the region Polaris is running in. I'm also totally fine with making it a required field to keep things clear and explicit.

XJDKC replied:
Hey @singhpk234, Polaris might be running in region A, while the remote catalog (like Glue) could be in region B.

dimas-b replied:
from my POV explicit region is preferable to assuming the region where Polaris runs (the latter may not be even inside AWS infra). Defaulting to the env.-level region setting is fine, though.

XJDKC replied:
Done!

adnanhemani reviewed (approved):
I have the same comments as @dimas-b and @singhpk234 - but this should be good for now!

XJDKC commented:
Hi folks, any other concerns about this spec change?

There is one remaining thing we haven't made a decision:
For non-STS based SigV4 authentication, how do we plan to support it? 

#### 1. **From env vars or server config**, e.g.:
* POLARIS_IAM_USER_AWS_ACCESS_KEY_ID
* POLARIS_IAM_USER_AWS_SECRET_ACCESS_KEY
* POLARIS_IAM_USER_ARN
In this case, `roleArn` would not be required.

#### 2. **Configured via the Polaris Management API: Stick to `SigV4AuthenticationParameters`**
If we stick with the existing `SigV4AuthenticationParameters` type, we could:
* Make roleArn optional
* Add `iamUserAwsAccessKeyId` and `iamUserAwsSecretAccessKey` as optional fields

#### 3. **Configured via the Polaris Management API: Add new auth type**

We could create a new type to distinguish clearly:
* New AuthenticationType enum: SIGV4_STS, SIGV4_STATIC_CREDS

#### 4. **Configured via the Polaris Management API: Add new auth types**

We could create a new sub type to distinguish clearly:
e.g. new subtype under SigV4AuthenticationParameters: STS, CREDS

Do you have any perference or is the current proposal good to go?

adnanhemani commented:
@XJDKC - sorry, I know I had previously approved this proposal but reading through some of the comments, I'm questioning something:

Why do we require a IAM User for Polaris? My understanding is that today, Polaris already has access to some IAM Role credentials (through the Default Credential Provider) - why do we need to introduce the new complexity of an IAM User? If we assume that Polaris has access to the IAM Role, then we no longer need the workaround for cross-account access as well.

Can you clarify on this bit?

On non-STS SigV4 authN, I don't really have a strong preference - and I don't think adopting any of the options constitutes this PR as a one-way door. So I'm okay to punt on it for now.

XJDKC commented:
> @XJDKC - sorry, I know I had previously approved this proposal but reading through some of the comments, I'm questioning something:
> 
> Why do we require a IAM User for Polaris? My understanding is that today, Polaris already has access to some IAM Role credentials (through the Default Credential Provider) - why do we need to introduce the new complexity of an IAM User? If we assume that Polaris has access to the IAM Role, then we no longer need the workaround for cross-account access as well.
> 
> Can you clarify on this bit?
> 
> On non-STS SigV4 authN, I don't really have a strong preference - and I don't think adopting any of the options constitutes this PR as a one-way door. So I'm okay to punt on it for now.

Hey @adnanhemani, there is an on-going discussion about this PR: https://lists.apache.org/thread/rlbxvw0xmzvlfm7pdh97bs3xvq7o8lmy

This thread also contains some details about it: https://github.com/apache/polaris/pull/1506#discussion_r2070930235

In short, to get an IAM Role credentials, polaris needs to use a long-lived credential (represents the polaris service identity) to assume the role (provided by polaris users) and get a temp credential. This is the solution provided by AWS IAM to grant permissions to AWS account B (polaris service provider) from AWS account A (polaris users).

dimas-b commented:
I believe the STS credentials concern is not specific to this PR. We can probably deal with that in another PR.

dimas-b commented on spec/polaris-management-service.yml:
```diff
@@ -938,6 +940,40 @@ components:
+          type: string
+          description: The service name to be used by the SigV4 protocol for signing requests, the default signing name is "execute-api" is if not provided
+          example: "glue"
+        userArn:
+          type: string
+          description: The aws user arn used to assume the aws role, this represents the polaris service itself
+          example: "arn:aws:iam::123456789001:user/polaris-service-user"
```
I believe `userArn` is not part of the Assume Role API call in STS... unless I'm missing something. So this parameter seems redundant in the connection config.

XJDKC replied:
This property is used to give the polaris service's `userArn` to polaris users, so that users can modify the trust relationship of the IAM role to allow polaris service to assume the role they provide.

Editing the trust relationship for an existing IAM role: https://docs.aws.amazon.com/directoryservice/latest/admin-guide/edit_trust.html

Storage config also has this property.

dimas-b replied:
Sorry, @XJDKC : Is it not the user who invokes the Admin API to save this config? The caller of the API must already know the value, right?

dimas-b replied:
> Storage config also has this property.

I believe this property is equally misleading in the storage config too :slightly_smiling_face:

dimas-b replied:
If the service owner needs to inform users about `userArn`, I believe we need to find another mechanism. Adding properties to configuration objects that are meant to affect Polaris behaviour, but have parameters not used by Polaris code is not a sound approach, I think.

dimas-b replied:
Side note: AFAIK, the trust policy does not have to be based on a "user" ARN, it could be based on a "role" ARN too.

XJDKC replied:
> Is it not the user who invokes the Admin API to save this config? The caller of the API must already know the value, right?

It's true if Apache Polaris is deployed in a controlled environment where users own both the AWS storage and the Polaris servers. But not the case if the the polaris service is provided by another vender, e.g. Snowflake, Dremio.

XJDKC replied:
> the trust policy does not have to be based on a "user" ARN, it could be based on a "role" ARN too.

Yes, you are right. But the fact is, we can't create long-lived creds for an IAM role. 

So if polaris wants to assume a user-privided role from a polaris service role, before that, polaris needs to get a temp creds for that polaris service role. But then how do we get those initial creds? By assuming that role from another set of long-lived credentials? That gets a bit weird and kind of defeats the purpose. 🙂

dimas-b replied:
> how about using serviceArn as the field name

This is still a value that it _not_ used by Polaris code. That, IMHO, makes it irrelevant to this config object and confusing.

I believe connection configuration in for cases when Polaris is a service needs to be clarified. 

What workflow do you have in mind? Something like this?
1. User contacts service provider to setup a Federated Catalog
2. Service provider invokes Admin API and saves a `SigV4AuthenticationParameters`
3. User accesses the Admin API and reads `SigV4AuthenticationParameters`
4. User extracts `serviceArn` from the output of the API call
5. User uses `serviceArn` to grant Polaris permission to perform AssumeRole in STS and access AWS resources.

... or is it something else?

If so, why couldn't the user learn about `serviceArn` at step 1?

dennishuo replied:
Right, that flow is indeed how the `StorageConfigInfo` works today for managed Polaris services that bind their own service-owned IAM User at the time of the `CreateCatalog` call; in general the binding of the IAM User to a StorageConfig (or a ConnectionConfig) may not be known until the `CreateCatalog` actually runs.

I think the problems with trying to do it at step 1 are:

- The binding might be dynamic and transactionally done performed during the CreateCatalog
- There's no separate API to express pre-planning for intending to create a Federated Catalog

It's true that if the service provider has very static base IAM Users or if the user interaction is very manual (e.g. if it's a small-scale/informal use case where an internal "user" DMs the "service provider" on Slack), then the userArn can be known already before creating the catalog.

Looks like there was some discussion about this already in https://github.com/apache/polaris/pull/1191

I know @XJDKC has proposed having a stronger delineation between user-provided and service-provided properties, and I agree it's worth cleaning up.

The "not used by Polaris" might just be a shortcoming of the current default implementations -- in a way we *should* be enforcing consistency of the userArn when setting up the `stsClient` in https://github.com/apache/polaris/blob/40aea3ac39ac485a76f3aaa930694eeab8ec14d8/service/common/src/main/java/org/apache/polaris/service/storage/StorageConfiguration.java#L63 


    default Supplier<StsClient> stsClientSupplier() {
      return Suppliers.memoize(
          () -> {
            StsClientBuilder stsClientBuilder = StsClient.builder();
            if (awsAccessKey().isPresent() && awsSecretKey().isPresent()) {
              LoggerFactory.getLogger(StorageConfiguration.class)
                  .warn("Using hard-coded AWS credentials - this is not recommended for production");
              StaticCredentialsProvider awsCredentialsProvider =
                  StaticCredentialsProvider.create(
                      AwsBasicCredentials.create(awsAccessKey().get(), awsSecretKey().get()));
              stsClientBuilder.credentialsProvider(awsCredentialsProvider);
            }
            return stsClientBuilder.build();
          });
    }


But the default implementation is just kept simple by basically taking a single root IAM User from the runtime environment that ends up being the userArn used implicitly for *all* StorageConfigurations in the service.

One step to expand on that architecture in a way that better illustrates how the fields actually should be used by the code would be if we allow something like setting per-realm root IAM Users or IAM Roles in the Quarkus config for being the "service identity" that will `assumeRole` on the *user*-specified catalog-level roleArn. Then the stsClient provider would be initially authenticating as the realm-specific "service identity" that matches what's in the Catalog's properties.

dennishuo replied:
Looks like the concept of "read-only" or "write-only" fields are actually first-class in OpenAPI: https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.2.md#fixed-fields-20

readOnly:

> Relevant only for Schema "properties" definitions. Declares the property as "read only". This means that it MAY be sent as part of a response but SHOULD NOT be sent as part of the request. If the property is marked as readOnly being true and is in the required list, the required will take effect on the response only. A property MUST NOT be marked as both readOnly and writeOnly being true. Default value is false.

writeOnly:

> Relevant only for Schema "properties" definitions. Declares the property as "write only". Therefore, it MAY be sent as part of a request but SHOULD NOT be sent as part of the response. If the property is marked as writeOnly being true and is in the required list, the required will take effect on the request only. A property MUST NOT be marked as both readOnly and writeOnly being true. Default value is false.


Maybe we should experiment with some of those additional annotations? Interestingly, `writeOnly` conveys well other patterns we've had such as inline secrets.

And `readOnly` would be fields used for communicating things decided by the server backend back to the user.

XJDKC replied:
Hey Dmitri and Dennis, thanks for the suggestions! Extracting the service-provided properties into a nested field and annotating it with `readOnly` both make a lot of sense to me. I'll update the spec to reflect these changes.

XJDKC replied:
Done!

dennishuo reviewed (approved):
LGTM given the consistency with StorageConfig, though I agree we should probably refine the way we convey "output-only" concepts like the `userArn`, and get agreement with other reviewers on how best to convey that distinction before merging this PR (maybe simply annotating with the openApi field `readOnly` is enough, or we could go further and move the readOnly fields into a separate tuple such as `vendorInfo` like you suggested elsewhere).

XJDKC commented:
Hi folks, I've updated the spec to extract the Polaris service identity info into a nested object.

Right now, I’ve added support for two types of service identities:
* **AWS_IAM_USER**: Polaris is configured with long-lived AWS credentials from an IAM user. It uses these credentials to assume the IAM role provided by Polaris users, then accesses the remote catalog using the subscoped temp credentials.
* **AWS_IAM_USER**: Polaris retrieves temporary credentials via [Amazon ECS container credentials](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-iam-roles.html) or [Amazon EC2 instance IAM role credentials](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html), then assumes the polaris user-provided IAM role to get new temporary credentials to access the remote catalog.

Looking ahead, we may introduce additional service identity types (e.g., GCP Service Account) and aim to unify the behavior between storage config and connection config for consistency.

cc: @dennishuo @dimas-b

dimas-b reviewed (commented):
LGTM after a quick glance. I'll take a more thorough read on Mon. Hope there's no rush :)

XJDKC commented:
Just wanted to give a warm heads-up on this PR: should we consider making serviceIdentityInfo a top-level property of `ConnectionConfigInfo` / `AuthenticationParameters` / `StorageConfigInfo`? Right now, I put it inside the `SigV4AuthenticationParameters`.

XJDKC commented on spec/polaris-management-service.yml:
```diff
@@ -938,6 +940,38 @@ components:
+          type: string
+          description: The service name to be used by the SigV4 protocol for signing requests, the default signing name is "execute-api" is if not provided
+          example: "glue"
+        serviceIdentity:
+          $ref: '#/components/schemas/ServiceIdentityInfo'
```
Hey Dmitri, I discussed the proposal of adding another set of management apis for the service identity info in my design doc (topic 3, proposal 3): [Apache Polaris Creds Management Proposal](https://docs.google.com/document/d/1MAW87DtyHWPPNIEkUCRVUKBGjhh5bPn0GbtV7fifm30/edit?usp=sharing)

Here is the pros and cons:
* Pros
  * Clean separation of concerns: identity info is no longer mixed into catalog or storage config
  * Keeps catalog schema simple and focused on user input
  * Easy to extend without touching catalog model or breaking clients
  * Works well for self-managed and SaaS deployments
* Cons
  * Users must make an additional API call to fetch identity info, it requires users to coordinate across two APIs to complete setup
  * Identity context is no longer colocated with the catalog entity, this assumed that polaris will use same identity for all the catalogs within the same realm
  * Incompatible with current behavior where identity fields are visible in the catalog response
  * Some service-managed fields like consentUrl are not static and cannot be precomputed globally. Instead, they depend on user-provided configuration (e.g., Azure tenant ID) and are generated after Polaris receives that info.

dimas-b replied:
Thanks for the analysis, @XJDKC !

The "Dynamic fields challenge" is very interesting. If I understand correctly, the user first creates an external catalog with some connection config details. The server processes that information and returns its identity information.

WDYT about making `ServiceIdentityInfo` a read-only sibling property to `connectionConfigInfo` in `ExternalCatalog`?

From my POV that approach still achieves separation of concerns between user-provided and server-provided data (server-provided in not embedded inside user-provided). At the same time, it allows for different server identities for each catalog.

XJDKC replied:
Hey @dimas-b, but there is another concern here: we may use different service identity to access different external services: e.g. 
> e.g. use SIGV4 auth to access Glue/Amazon API Gateway (host polaris behind the API Gateway), but the table is stored in azure blob (or s3 comp storage).

dimas-b replied:
That fits into my proposal (previous comment) does it not?

adutra replied:
I don't think a "read-only sibling property to `connectionConfigInfo`" would work, would it?

How about:

```json5
{
  "name": "aws-catalog",
  "connectionConfigInfo": {
    "authenticationParameters": { /* ... */ },
    "serviceIdentity": { /* ... */ }
  },
  "storageConfigInfo": {
    "serviceIdentity": { /* ... */ }
  }
}
```

dimas-b replied:
Good point, @adutra . That matches my idea, but I guess I was not very careful with exact object placement in my suggestion :sweat_smile:

XJDKC replied:
Yeah, that matches my idea as well!

This isn't directly related to the spec changes, but I want to clarify how Polaris retrieves connection credentials at runtime.

The `serviceIdentity` field is intended to be surfaced to Polaris users, for example, so they can update their IAM role's trust policy to allow Polaris to assume it. This is the user-facing identity info.

But in a multi-tenant Polaris deployment, each catalog may be assigned a different service identity behind the scenes. Polaris needs to know where to retrieve the actual credentials tied to that identity, e.g., AWS user credentials stored in Secrets Manager.

So how does Polaris know where to find those credentials?

For storage config, we already follow this pattern. There’s a hidden field in `internalProperties` called [storage_integration_identifier](https://github.com/apache/polaris/blob/apache-polaris-0.10.0-beta-incubating-rc2/polaris-core/src/main/java/org/apache/polaris/core/entity/PolarisEntityConstants.java#L49-L50), which points to the service identity's credentials. Polaris loads a `PolarisStorageIntegration` based on this value and uses it to retrieve credentials and generate subscoped credentials as needed: [example here](https://github.com/apache/polaris/blob/apache-polaris-0.10.0-beta-incubating-rc2/polaris-core/src/main/java/org/apache/polaris/core/persistence/transactional/TransactionalMetaStoreManagerImpl.java#L2002-L2012).

**Option 1: We could apply the same pattern for connection config:**
* Use serviceIdentity to expose the Polaris identity to the user
* Use a hidden field (e.g., connection_integration_identifier) to locate the actual credentials Polaris needs to access external services

At runtime, Polaris would use that identifier to fetch the real credential (e.g. from a secret manager), then assume the user-provided IAM role and generate temporary, subscoped credentials.

**Option 2**: Use a reference field in the persistence model `ServiceIdentityDpo`
Alternative, instead of introducing another hidden sibling field like `connection_integration_identifier`, we could embed a reference directly inside the service identity model (`ServiceIdentityDpo`), similar to how `UserSecretReference` ([OAuthClientCredentialsParametersDpo.java#L64-L65](https://github.com/apache/polaris/blob/apache-polaris-0.10.0-beta-incubating-rc2/polaris-core/src/main/java/org/apache/polaris/core/connection/OAuthClientCredentialsParametersDpo.java#L64-L65)) works.

For example, we could add a `serviceInfoAuthLocatorUrn` field to the internal `ServiceIdentityDpo`. **The spec would remain unchanged, but the persistence model would carry the locator used to find the backing credentials.**

```
{
  "name": "aws-catalog",
  "connectionConfigInfo": {
    "authenticationParameters": { /* ... */ },
    "serviceIdentity": {
      "type": "AWS_IAM",
      "serviceArn": "arn:aws:iam::111122223333:user/polaris-service-user",
      "serviceInfoAuthLocatorUrn": "urn:polaris-service-identities:realm-id:catalog-id:connection"
    }
  },
  "storageConfigInfo": {
    "serviceIdentity": { /* ... */ }
  }
}
```

WDYT?
cc: @dimas-b @adutra @dennishuo

dimas-b replied:
I assume credentials referenced by `serviceInfoAuthLocatorUrn` are provided by the end user, right? In that case does it not contradict having a `serviceArn`?

In other words, if the user provide connection credentials, the "service" acts not by itself, but on behalf of the user. It will assume the identity associated with the use-provided credentials. In this case we probably do not need `serviceIdentity` at all :thinking:  ... or did I miss something?

I think `serviceIdentity` is relevant only when the service authenticates itself using service-provided (pre-configured) credentials (such as workload identity). WDYT?

XJDKC replied:
Nope, the `serviceInfoAuthLocatorUrn` is not provided by end user. All the properties in `serviceIdentity` should be provided by polaris itself. The different is:
`serviceArn` is surfaced to polaris users so that they can add it (IAM user arn) to their trust relationship of IAM role.
`serviceInfoAuthLocatorUrn`: is used to look up the IAM user's credential from the secret manager, it's owned by polaris services, polaris will assign an IAM user for a specific Catalog Entity.

dimas-b replied:
In this case should `serviceInfoAuthLocatorUrn` be exposed in the user-facing API?

As I understand, it does not expose actual credentials, just a reference to them, but this information seems to be irrelevant for users of the API :thinking: 

We could still keep it in `ServiceIdentityDpo` in the Polaris Server, just not expose in API. WDYT?

XJDKC replied:
Yes, that's my plan, `serviceInfoAuthLocatorUrn` will only exist in `ServiceIdentityDpo` and won't be included in the `ServiceIdentityInfo` in the API response. Since it's our internal implementation details, only polaris needs this info.

adutra replied:
Thanks for the detailed explanations, much appreciated!

Given that the location URN is not surfaced to users, I think that this option (2) would have my preference. But we should maybe look into migrating the storage config to use the same system eventually.

XJDKC replied:
Yes, in the long term, we should use the same system for both storage config and connection config.

I discussed this in the proposal, feel free to take a look if you're interested (though it's a bit outdated based on the latest discussions, and I'll update it soon).: [Apache Polaris Creds Management Proposal](https://docs.google.com/document/d/1MAW87DtyHWPPNIEkUCRVUKBGjhh5bPn0GbtV7fifm30/edit?usp=sharing)

XJDKC commented on spec/polaris-management-service.yml:
```diff
@@ -1023,6 +1057,49 @@ components:
+        identityType:
+          type: string
+          enum:
+            - AWS_IAM_USER
+            - AWS_IAM_ROLE
```
Makes sense, I will only introduce a new type `AWS_IAM` to capture these two!

XJDKC replied:
Done!

dimas-b reviewed (approved):
spec change LGTM. Thanks @XJDKC !

Let's keep the PR open for a few more days in case other reviews have more comments.

adutra commented on spec/polaris-management-service.yml:
```diff
@@ -868,6 +868,8 @@ components:
           description: URI to the remote catalog service
         authenticationParameters:
           $ref: "#/components/schemas/AuthenticationParameters"
+        serviceIdentity:
```
Shouldn't this field be marked as read-only?

XJDKC replied:
The ref schema is read-only already, so we should be good.

dennishuo reviewed (approved):
LGTM

XJDKC commented:
Thanks everyone for reviewing this PR! I really appreciate the thoughtful comments which helped me think more thoroughly about the problem and refine the proposal to cover more use cases.

I'll leave this PR open until Monday to allow any final feedback. After that, I'll start implementing based on the discussions and the updated spec.

The initial implementation will focus on SigV4 Auth support for Catalog Federation. Longer term, I hope we can move toward a unified model for how Polaris handles both storage and connection credential management across deployment modes.

dimas-b commented:
I propose to merge on May 20
